### PR TITLE
fix(iot-dev): Fix semaphore usage within InternalClient

### PR DESCRIPTION
**Pull request by timtay-microsoft**
This PR fixes the InternalClient layer's semaphore usage so that each semaphore will never be released within a method unless it was grabbed. The new pattern that I'm implementing looks like the below:

```csharp
public async Task Foo(CancellationToken cancellationToken)
{
    cancellationToken.ThrowIfCancellationRequested();
    await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);

    try
    {
        // Perform all operations that the semaphore protects like updating local variable states, subscribing over AMQP/MQTT, etc
        // Any number of timeouts, IOExceptions, etc may be thrown here, but the semaphore will be released regardless
    }
    finally
    {
        _methodsDictionarySemaphore.Release();
    }
}
```

With this approach, the finally block only executes if the semaphore is grabbed, and it will be released no matter what exceptions are encountered after grabbing it. Additionally, if the provided cancellation token expires before or during the wait for grabbing the semaphore, an exception will be thrown, and the semaphore won't released since these timeouts don't occur in the try/finally block.

#1821 for additional context.

### CODE REVIEW

File: iothub/device/src/InternalClient.cs [hunk 1]
```diff
@@ -819,20 +819,7 @@ public async Task SetMethodHandlerAsync(string methodName, MethodCallback method
             }
             finally
             {
-                try
-                {
-                    _methodsDictionarySemaphore.Release();
-                }
-                catch (SemaphoreFullException)
```
Comment (timtay-microsoft) [inline iothub/device/src/InternalClient.cs:826]: This catch was part of an earlier attempt at fixing a similar bug, but it won't be needed anymore since this finally block will only be reached if the semaphore was grabbed

File: iothub/device/src/InternalClient.cs [hunk 2]
```diff
@@ -1835,10 +1753,11 @@ public async Task SetMessageHandlerAsync(MessageHandler messageHandler, object u
                 Logging.Enter(this, messageHandler, userContext, nameof(SetMessageHandlerAsync));
             }
 
+            cancellationToken.ThrowIfCancellationRequested();
+            await _moduleReceiveMessageSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
             try
             {
-                cancellationToken.ThrowIfCancellationRequested();
-                await _moduleReceiveMessageSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
```
Comment (timtay-microsoft) [inline iothub/device/src/InternalClient.cs:1841]: As per the model I described in the description of this PR, grabbing the semaphore and checking the cancellation token must be outside of the try/finally block to ensure that we don't try to release a semaphore that we never grabbed.

File: iothub/device/src/InternalClient.cs [hunk 3]
```diff
@@ -1468,11 +1402,7 @@ private async Task OnDeviceMessageReceivedAsync(Message message)
 
             try
             {
-                // Wait to acquire the _deviceReceiveMessageSemaphore. This ensures that we get a reference to the latest callback set on the device client.
-                await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
                 callbackContextTuple = _deviceReceiveMessageCallback;
-                _deviceReceiveMessageSemaphore.Release();
```
Comment (timtay-microsoft) [inline iothub/device/src/InternalClient.cs:1474]: This change is a bit orthogonal, but there was no reason to use a semaphore here. It reads from a shared resource, but it doesn't write to the shared resource so there is no concurrency risk

File: iothub/device/src/InternalClient.cs [hunk 4]
```diff
@@ -1468,11 +1402,7 @@ private async Task OnDeviceMessageReceivedAsync(Message message)
 
             try
             {
-                // Wait to acquire the _deviceReceiveMessageSemaphore. This ensures that we get a reference to the latest callback set on the device client.
-                await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
```
Comment (abhipsaMisra) [inline iothub/device/src/InternalClient.cs:1472]: This lock shouldn't be removed. We want to ensure that the callback is invoked on the latest supplied handler.

File: iothub/device/src/InternalClient.cs (hunk 3 above)
Comment (abhipsaMisra) [reply iothub/device/src/InternalClient.cs:1474]: What about the scenario where a user supplies a different callback right before the client was about to actually invoke it?

File: iothub/device/src/InternalClient.cs (hunk 4 above)
Comment (timtay-microsoft) [reply iothub/device/src/InternalClient.cs:1472]: Even with this lock, the handler could get updated immediately after this value is read. I can remove this for now since it isn't critical for fixing the github issue, but I don't see the value of leaving it.

File: iothub/device/src/InternalClient.cs (hunk 3 above)
Comment (timtay-microsoft) [reply iothub/device/src/InternalClient.cs:1474]: The callback can be updated immediately after this value read even with this lock. Even with the semaphore, there is no guarantee that the read callback will still be valid after a millisecond. Like I mentioned on the other comment, I can remove this change since it isn't critical to fix the github issue in question, but we should still consider removing it at some point

File: iothub/device/src/InternalClient.cs (hunk 4 above)
Comment (abhipsaMisra) [reply iothub/device/src/InternalClient.cs:1472]: I agree, it could, but since we invoke it immediately after, it does reduce the scope of us invoking the handler on an updated callback. Also, you could release the semaphore after invoking the callback, that way a user won't be able to update the callback in the interim.

File: iothub/device/src/InternalClient.cs [hunk 5]
```diff
@@ -1009,22 +984,8 @@ internal async Task OnMethodCalledAsync(MethodRequestInternal methodRequestInter
...
-                    // this method is canceled while waiting to grab the semaphore, then this semaphore.release
-                    // will throw this SemaphoreFullException since it was never grabbed in the first place
-                    if (Logging.IsEnabled)
-                    {
-                        Logging.Info(this, "SemaphoreFullException thrown while releasing methods dictionary semaphore," +
-                            " but will be ignored since that means the semaphore is available for other threads to grab again anyways");
-                    }
-                }
+                methodResponseInternal?.Dispose();
+                _methodsDictionarySemaphore.Release();
```
Comment (drwill-ms) [inline iothub/device/src/InternalClient.cs:988]: Something to consider is that if the call to Dispose throws, then we won't release the semaphore

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/InternalClient.cs (hunk 3 above)
Comment (jamdavi) [reply iothub/device/src/InternalClient.cs:1474]: Looking at this more If we're trying to guarantee that the callback is guaranteed we'd need to do the following. If we don't care then Tim is correct the Semaphore is not helping here.

1. Initialize a new tuple so we get the correct references
2. Move the lock to after the callback finished.

```
            try
            {
                // Wait to acquire the _deviceReceiveMessageSemaphore. This ensures that we get a reference to the latest callback set on the device client.
                await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
                Tuple<ReceiveMessageCallback, object> callbackContextTuple = new Tuple<ReceiveMessageCallback, object>(_deviceReceiveMessageCallback.Item1, _deviceReceiveMessageCallback.Item2);
                _deviceReceiveMessageSemaphore.Release();

                await (callbackContextTuple?.Item1?.Invoke(message, callbackContextTuple.Item2)
                    ?? TaskHelpers.CompletedTask).ConfigureAwait(false);
            }
            finally
            {
                if (Logging.IsEnabled)
                {
                    Logging.Exit(this, message, nameof(OnDeviceMessageReceivedAsync));
                }
            }
```

File: iothub/device/src/InternalClient.cs (hunk 4 above)
Comment (timtay-microsoft) [reply iothub/device/src/InternalClient.cs:1472]: I'll move the semaphore here to fully wrap the scope of the read and invoke of the callback then, sure

File: iothub/device/src/InternalClient.cs (hunk 5 above)
Comment (timtay-microsoft) [reply iothub/device/src/InternalClient.cs:988]: Good point. This will require a small rework to take that into account

**Review by bikamani — APPROVED** [review]

**Comment by bikamani** [comment]
> This PR fixes the InternalClient layer's semaphore usage so that each semaphore will never be released within a method unless it was grabbed. The new pattern that I'm implementing looks like the below:
> 
> ```cs
> public async Task Foo(CancellationToken cancellationToken)
> {
>     cancellationToken.ThrowIfCancellationRequested();
>     await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
> 
>     try
>     {
>         // Perform all operations that the semaphore protects like updating local variable states, subscribing over AMQP/MQTT, etc
>         // Any number of timeouts, IOExceptions, etc may be thrown here, but the semaphore will be released regardless
>     }
>     finally
>     {
>         _methodsDictionarySemaphore.Release();
>     }
> }
> ```
> 
> With this approach, the finally block only executes if the semaphore is grabbed, and it will be released no matter what exceptions are encountered after grabbing it. Additionally, if the provided cancellation token expires before or during the wait for grabbing the semaphore, an exception will be thrown, and the semaphore won't released since these timeouts don't occur in the try/finally block.
> 
> #1821 for additional context.

Wondering is there a way to add a test for this scenario for future protection?

**Comment by timtay-microsoft** [comment]
> Wondering is there a way to add a test for this scenario for future protection?

I don't know of any way to test that scenario since it relies on the randomness of threading.

File: iothub/device/src/InternalClient.cs (hunk 3 above)
Comment (timtay-microsoft) [reply iothub/device/src/InternalClient.cs:1474]: Yep. Since I'm getting some pushback on removing the semaphore grab here, I did what you suggested above, but without declaring the tuple. Now it is 

```csharp
            await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
            try
            {
                await (_deviceReceiveMessageCallback?.Item1?.Invoke(message, _deviceReceiveMessageCallback.Item2)
                    ?? TaskHelpers.CompletedTask).ConfigureAwait(false);
            }
            finally
            {
                _deviceReceiveMessageSemaphore.Release();
             
                if (Logging.IsEnabled)
                {
                    Logging.Exit(this, message, nameof(OnDeviceMessageReceivedAsync));
                }
            }
```

Comment (timtay-microsoft) [reply iothub/device/src/InternalClient.cs:1474]: Now the lock is a bit more useful since it makes sure that the two calls to get values from the callback (.Item1 and .Item2) will come from the same callback instance. Without the lock, there would be a chance that the callback value is set inbetween the retrieval of item1 and item2

**Review by abhipsaMisra — APPROVED** [review]
lgtm!

File: iothub/device/src/InternalClient.cs [hunk 6]
```diff
@@ -1464,20 +1405,16 @@ private async Task OnDeviceMessageReceivedAsync(Message message)
...
+            await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
             try
             {
-                // Wait to acquire the _deviceReceiveMessageSemaphore. This ensures that we get a reference to the latest callback set on the device client.
-                await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
-                callbackContextTuple = _deviceReceiveMessageCallback;
-                _deviceReceiveMessageSemaphore.Release();
-
-                await (callbackContextTuple?.Item1?.Invoke(message, callbackContextTuple.Item2)
+                await (_deviceReceiveMessageCallback?.Item1?.Invoke(message, _deviceReceiveMessageCallback.Item2)
```
Comment (drwill-ms) [inline iothub/device/src/InternalClient.cs:1411]: I think this would be easier to read if you just checked for null

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/InternalClient.cs (hunk 3 above)
Comment (jamdavi) [reply iothub/device/src/InternalClient.cs:1474]: Not sure I like this since we will block the semaphore until method is complete. Let's get a reference to the two items from the Tuple during the Semaphore operation in the Semaphore context and then invoke the callback after we've released the lock.

Comment (jamdavi) [reply iothub/device/src/InternalClient.cs:1474]: (code written not compiled)

```C#
await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
var item1 = _deviceReceiveMessageCallback?.Item1;
var item2 = _deviceReceiveMessageCallback?.Item2;
_deviceReceiveMessageSemaphore.Release();
            try
            {
                await (item1?.Invoke(message, item2)
                    ?? TaskHelpers.CompletedTask).ConfigureAwait(false);
            }
            finally
            {            
                if (Logging.IsEnabled)
                {
                    Logging.Exit(this, message, nameof(OnDeviceMessageReceivedAsync));
                }
            }
```

Comment (timtay-microsoft) [reply iothub/device/src/InternalClient.cs:1474]: Good point. Callbacks can take a pretty variable amount of time

Comment (abhipsaMisra) [reply iothub/device/src/InternalClient.cs:1474]: Wait, how is this different from the code that we had before?

**Review by jamdavi — APPROVED** [review]

File: iothub/device/src/InternalClient.cs (hunk 3 above)
Comment (abhipsaMisra) [reply iothub/device/src/InternalClient.cs:1474]: There isn't a need for us to await the completion of this user supplied callback, we can simply invoke it and then release the semaphore.

Comment (timtay-microsoft) [reply iothub/device/src/InternalClient.cs:1474]: As of now, the code behaves the same way as it used to. I just refactored it a bit for readability
